File: ruby/soaparse.rb
def soaparse(soa_string)
    soa_parts = soa_string.split(' ')

    # remove trailing periods from mname, rname
    mname = soa_parts[0][...-1]
    rname = soa_parts[1][...-1]

    parsed_email_beginning = rname.dup # .dup creates copy of rname to avoid mutability on original
    parsed_email_beginning["."] = "@" # replace first instance of period

    # replace second from last instance of period
    second_from_last_index = second_from_last(rname, ".")
    parsed_email_end = rname[...second_from_last_index] + "@" + rname[second_from_last_index + 1..]

    serial = soa_parts[2]
    refresh = soa_parts[3]
    retry_time = soa_parts[4] # retry is keyword in ruby
    expire = soa_parts[5]
    minimum = soa_parts[6]

    return {
        mname: mname,
        rname: rname,
        parsed_email_beginning: parsed_email_beginning,
        parsed_email_end: parsed_email_end,
        serial: serial,
        refresh: refresh,
        retry: retry_time,
        expire: expire,
        minimum: minimum
    }
end

# return index of the second to last instance of a char in a string
def second_from_last(string, char)
    return string.rindex(char, string.rindex(char) - 1)
end

File: clojure/soaparse.clj
(require '[clojure.string :as str])

(defn strip-period [str]
    (str/replace str #"\.$" "")
)

(defn replace-second-from-last [string char replacement]
    (def end_index
        (str/last-index-of string char)
    )

    (def second_last_index
        (str/last-index-of (subs string 0 end_index) char)
    )

    (str
        (subs string 0 second_last_index) replacement (subs string (+ 1 second_last_index))
    )
)

; "ns.example.com. email.example.com. 20200101 60 3 2 1"
(defn soaparse [soa_string]
    (def soa_parts (str/split soa_string #" "))

    ; trim period from mname, rname
    (def mname (strip-period (first soa_parts)))
    (def rname (strip-period (nth soa_parts 1)))

    (def parsed_email_beginning (str/replace-first rname "." "@"))
    (def parsed_email_end (replace-second-from-last rname "." "@"))

    (def serial (nth soa_parts 2)) ; serial
    (def refresh (nth soa_parts 3)) ; refresh
    (def retry (nth soa_parts 4)) ; retry
    (def expire (nth soa_parts 5)) ; expire
    (def minimum (nth soa_parts 6)) ; minimum

    ; return hash map of values
    (hash-map
        "mname" mname
        "rname" rname
        "parsed_email_beginning" parsed_email_beginning
        "parsed_email_end" parsed_email_end
        "serial" serial
        "refresh" refresh
        "retry" retry
        "expire" expire
        "minimum" minimum
    )
)

(soaparse "ns.example.com. email.example.com. 20200101 60 3 2 1")
